hibernate-jcache/src/main/java/org/hibernate/cache/jcache/access/AbstractReadWriteRegionAccessStrategy.java

```diff
@@ -0,0 +1,404 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.cache.jcache.access;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.hibernate.cache.CacheException;
+import org.hibernate.cache.jcache.JCacheMessageLogger;
+import org.hibernate.cache.jcache.JCacheTransactionalDataRegion;
+import org.hibernate.cache.spi.access.SoftLock;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.jboss.logging.Logger;
+
+/**
+ * @author Alex Snaps
+ */
+abstract class AbstractReadWriteRegionAccessStrategy<R extends JCacheTransactionalDataRegion> {
+
+	private static final JCacheMessageLogger LOG = Logger.getMessageLogger(
+			JCacheMessageLogger.class,
+			AbstractReadWriteRegionAccessStrategy.class.getName()
+	);
+
+	protected final R region;
+	protected final Comparator versionComparator;
+	private final UUID uuid = UUID.randomUUID();
+	private final AtomicLong nextLockId = new AtomicLong();
+	private final AtomicLong nextItemId = new AtomicLong();
+
+	public AbstractReadWriteRegionAccessStrategy(R region) {
+		this.versionComparator = region.getCacheDataDescription().getVersionComparator();
+		this.region = region;
+	}
+
+	public R getRegion() {
+		return region;
+	}
+
+	public Object get(SharedSessionContractImplementor session, Object key, long txTimestamp) throws CacheException {
+		final Lockable item = (Lockable) region.get( key );
+
+		final boolean readable = item != null && item.isReadable( txTimestamp );
+		if ( readable ) {
+			return item.getValue();
+		}
+		else {
+			return null;
+		}
+	}
+
+	public boolean putFromLoad(SharedSessionContractImplementor session, Object key, Object value, long txTimestamp, Object version) throws CacheException {
+		while (true) {
+			Lockable item = (Lockable) region.get( key );
+
+			if (item == null) {
+				/*
+				 * If the item is null due a softlock being evicted... then this
+				 * is wrong, the in-doubt soft-lock could get replaced with the
+				 * old value.  All that can be done from a JCache perspective is
+				 * to log a warning.
+				 */
+				if (region.putIfAbsent( key, new Item( value, version, txTimestamp, nextItemId() ))) {
+					return true;
+				}
+			}
+			else if (item.isWriteable( txTimestamp, version, versionComparator )) {
+				if (region.replace( key, item, new Item( value, version, txTimestamp, nextItemId() ))) {
+					return true;
+				}
+			}
+			else {
+				return false;
+			}
+		}
+	}
+
+	public boolean putFromLoad(SharedSessionContractImplementor session, Object key, Object value, long txTimestamp, Object version, boolean minimalPutOverride)
+			throws CacheException {
+		return putFromLoad( session, key, value, txTimestamp, version );
```

Comment:
If `minimalPutOverride = false`, you should do region.putIfAbsent without reading the value ahead, I think